exonum/src/proto/schema/exonum/runtime.proto

```diff
@@ -66,34 +66,53 @@ message GenesisConfig {
   repeated InstanceInitParams builtin_instances = 3;
 }
 
-// The kind of ExecutionError.
+// The type of ExecutionError.
+// Note that ErrorKind doesn't communicate the source of the error, for this
+// see the 'runtime_id' and 'call_site' fields of ExecutionError.
 enum ErrorKind {
+  // An unexpected error which does not have a specific cause.
+  // The description of the error is the only source of information about this kind of errors.
   UNEXPECTED = 0;
+  // An error specific to the core. See core error codes for details.
   CORE = 1;
+  // An error specific to the certain runtime. See error codes of corresponding runtime for details.
   RUNTIME = 2;
+  // An error specific to the certain service. See error codes of corresponding service for details.
   SERVICE = 3;
+  // A common error which can occur in different contexts. See common error codes for details.
   COMMON = 4;
 }
 
 // Result of unsuccessful runtime execution.
+// ExecutionError message provides the information about the source of the error.
+// The source of error is determined as following:
+// - If both runtime ID and call site are set, then error originates in the service code.
+// - If runtime ID is set, and call site is not set, then error originates in the runtime code.
+// - If none of runtime ID and call site is set, then error originates in the core code.
+// Option with set call site and unset runtime ID is not valid, receiving a message
+// with such a combination means receiving a malformed message.
 message ExecutionError {
-  // The kind of error that indicates in which module the error occurred.
+  // The kind of error that indicates its type.
   ErrorKind kind = 1;
   // User defined error code that can have different meanings for the different
   // error kinds.
   uint32 code = 2;
   // Optional description which doesn't affect `object_hash`.
   string description = 3;
 
+  // Runtime ID will be set if the error originates in code related to the certain runtime.
   oneof runtime {
-    // Runtime identifier associated with the error.
+    // Identifier of runtime in which error originates.
     uint32 runtime_id = 4;
-    // The runtime associated with the error is unknown.
+    // The error does not originate in the runtime code.
```

Comment:
This seems incorrect; at least this was not my intent when I designed these data structures in #1585. If `runtime_id` is set, this means there is a runtime responsible for processing the erroneous call. The error may originate either directly in the runtime, or in the service, depending on the *error kind*.

Similarly, if `call_site` is set, the error does not necessarily originate in the service code. For example, if there is no interface, a corresponding `CommonError` will be raised (at least in the Rust runtime). In other words, both error kind and error location are needed to fully interpret the error (e.g., service error codes don't tell much if we don't know which service they belong to - which information `call_site` provides).